### DISCUSSION TITLE
Passing arguments to useFetcher

### DISCUSSION THREAD

**Discussion by MaxwellGover** [remix-run/remix#4670] [Q&A]
Hello! New to the Remix framework. I have a component which I don't need a route for called `Teams`. Reading the docs I came across `useFetcher` which seems to be what I'm looking for? I'm trying to make a GET request using the `teamId` that is passed into my component. What is the best way to pass arguments into my api calls using Remix? My `Teams` component is also in a components folder which I don't know if that is ok or if it needs to be in the route folder? Appreciate any feedback. Thanks!

```
import { useEffect } from "react"
import { useFetcher } from "@remix-run/react";

export async function action() {
  const res = await fetch(
    `https://stats.nba.com/stats/scoreboardv2?DayOffset=0&TeamId=${teamId}`, {
    method: 'GET'
  })

  const json = await res.json();
  
  return { teamInfo: json }
}

export default function TeamInfo({ teamId }) {
  const fetcher = useFetcher();
  
  return (
    <div>
      TEAM
    </div>
  )
}
```

**Comment by marwan38** [comment]
params, url search params, post data.. whatever fits your user case.

From your example it looks like you should be using a dynamic route param like `/team/:teamId`

**Comment by sergiodxa** [comment]
> I'm trying to make a GET request

GET requests are handled by loaders, so your code should export a loader function and not an action.

After that, the way to send data to a loader is using the URL, either with route params or search params.

```ts
import { useEffect } from "react"
import { useFetcher } from "@remix-run/react";
import type { LoaderArgs } from "@remix-run/node";

export async function loader({ request, params }: LoaderArgs) {
  let { searchParams } = new URL(request.url)
  
  // get dayOffset from search params, default to 0 in case it's missing
  let dayOffset = Number(searchParams.get("dayOffset") ?? 0)

  // get teamId from route params, this means the route file exporting the loader must include
  // `$teamId` in the name
  let teamId = params.teamId

  // validate teamId exists, because search params can be string or undefined
  if (!teamId) return json({ error: "missing teamId" }, 400)

  const res = await fetch(
    `https://stats.nba.com/stats/scoreboardv2?DayOffset=${dayOffset}&TeamId=${teamId}`, {
    method: 'GET'
  })

  const json = await res.json();
  
  return json({ teamInfo: json })
}

export default function TeamInfo({ teamId }) {
  const fetcher = useFetcher();

  // this be undefined initialize, after you submit fetcher.Form it will be the data from the loader
  console.log(fetcher.data)
  
  return (
    // the action is where the form send a request, it's called action but it works for loaders too
    // the method by default is get so you can avoid setting it
    <fetcher.Form action="/the/url/of/your/loader" method="get">
      // the name here will be the name in the search params
      <input type="number" name="dayOffset" />
      <button>Submit</button>
    </fetcher.Form>
  )
}
```

> My Teams component is also in a components folder which I don't know if that is ok or if it needs to be in the route folder?

Your component can be anywhere, but the loader or action functions need to be exported from a route to work. You can use [resource routes](https://remix.run/docs/en/v1/guides/resource-routes) to export only a loader or action, essentially creating an [API route](https://remix.run/docs/en/v1/guides/api-routes).

---

In the example I used `fetcher.Form`, this is great if the fetch needs to happen after the user clicks a button so you can just render the Form, a button inside and some inputs and that's it, but you can use it with an effect if you need it immediately

```ts
useEffect(() => {
  fetcher.submit(
    // this replace the inputs, so here are your search params, with POST this is your formData
    { dayOffset: someValueHereYouGotSomehow },
    { action: "/the/url/of/your/loader", method: "get" }
  )
}, [someValueHereYouGotSomehow])
```

But in general, I recommend you to load data from a route loader if you need it to render a component.
